### ISSUE TITLE
Variable that self references causes deploy to hang

### ISSUE THREAD

**Issue by apsoto** [serverless/serverless#7807]
If you mistakenly have a variable reference itself it gets stuck in what appears to be an infinite loop.  I've waited many minutes, and the deploy process provides no feedback nor does it error out.

An example I mistakenly did due to a bad cut and paste in my serverless.yml:

```
provider:
  environment:
    SOME_VARIABLE: ${self:provider.environment.SOME_VARIABLE}
```

`sls deploy -v` hangs with no output

Other mention: https://github.com/serverless/serverless/issues/4687#issuecomment-637741810

<details>
<summary><code>serverless.yml</code></summary>

```yaml
service: test
provider:
  name: aws
  runtime: nodejs12.x
  environment:
    SOME_VARIABLE: ${self:provider.environment.SOME_VARIABLE}
functions:
  hello:
    handler: handler.hello

```

</details>

<details>
<summary><b><code>REPLACE WITH FULL COMMAND NAME</code> output</b></summary>

```
sls deploy -v
```

</details>

<b>Installed version</b>

```
Framework Core: 1.72.0
Plugin: 3.6.13
SDK: 2.3.1
Components: 2.30.12
```

**Comment by medikoo** [comment]
@apsoto thanks for reporting. It's definitely a bug, that at some point should be fixed.

We will possibly rewrite variables resolution logic at some point in time (we cover that here: https://github.com/serverless/serverless/issues/7207 ) and then we'll ensure that this issue is then addressed.

Otherwise if there's some quick fix that can be done just to address this issue, PR is definitely welcome

**Comment by medikoo** [comment]
Fixed with #8987
